### PR TITLE
nixos/shells: Avoid overriding the environment for other child shells

### PR DESCRIPTION
**Pull request by oxij**
###### Motivation for this change

An edited version of #44720.

###### Things done

- [X] It boots.
- [X] bash and zsh in tty, pty and over ssh work.
- [X] Checked fish.

/cc @lnl7 @pxc @rnhmjoj from git-blame

### CODE REVIEW

**Comment by Mic92** [comment]
What if someone wants to `. /etc/profile` after upgrading?

File: nixos/modules/config/shells-environment.nix [hunk 1]
```diff
@@ -162,15 +162,24 @@ in
         /bin/sh
       '';
 
+    # For resetting environment with `. /etc/set-environment` when needed
+    # and discoverability (see motivation of #30418).
+    environment.etc."set-environment".source = config.system.build.setEnvironment;
```
Comment (Mic92) [inline nixos/modules/config/shells-environment.nix]: Ah I see. This should be documented for sure!

**Comment by oxij** [comment]
> What if someone wants to `. /etc/profile` after upgrading?

Everything would stay as it was before this PR as it sets `__ETC_PROFILE_SOURCED=1`.

**Comment by oxij** [comment]
ping?

I agree that the new `/etc` thing should be documented somewhere, but I don't see a good place in the manual where to put it, and I'm not feeling like making one right now.

**Comment by hedning** [comment]
This does fix a rather unfortunate bug with `nix run`, where it will reset PATH if the parent shell isn't bash (or in a typical nixos graphical session which source `/etc/profile`).

**Comment by rnhmjoj** [comment]
@oxij I'll test this with fish soon. If I understand correctly this should also fix #18946.

**Comment by oxij** [comment]
Yes, you're right, it should.

**Comment by rnhmjoj** [comment]
Everything seems to be working correctly, no problem so far.

**Comment by oxij** [comment]
If fish works, then LGTM.

Can somebody merge this? Please also backport to 18.09.

This fixes #18946, #33219 and unblocks #44497.

**Comment by oxij** [comment]
ping @Mic92 (you commented above), @xeji (you merged a bunch of nixos things recently)

**Review by xeji — APPROVED** [review]
LGTM

**Comment by xeji** [comment]
backported in e981546..23da995a02b

**Comment by oxij** [comment]
@xeji Thanks! Can I also ask you to do the honors of closing #18946, #33219, and #46091?

**Comment by xeji** [comment]
done

**Comment by oxij** [comment]
@xeji Huge thanks!

Thanks to @hedning for the persistence on the issue and to everyone else here for testing.

Cheers!

**Pull request merged**